src/basic_operations.jl

```diff
@@ -7,27 +7,28 @@
-@inline Base.:+{order, dim, T}(S1::SymmetricTensor{order, dim, T}, S2::SymmetricTensor{order, dim, T}) = map(+, S1, S2)
-@inline Base.:+{order, dim}(S1::AbstractTensor{order, dim}, S2::AbstractTensor{order, dim}) = +(promote(S1, S2)...)
-@inline Base.:-{order, dim, T}(S1::Tensor{order, dim, T}, S2::Tensor{order, dim, T}) = map(-, S1, S2)
-@inline Base.:-{order, dim, T}(S1::SymmetricTensor{order, dim, T}, S2::SymmetricTensor{order, dim, T}) = map(-, S1, S2)
-@inline Base.:-{order, dim}(S1::AbstractTensor{order, dim}, S2::AbstractTensor{order, dim}) = -(promote(S1, S2)...)
+@inline Base.:+{order, dim}(S1::Tensor{order, dim}, S2::Tensor{order, dim}) = map(+, S1, S2)
+@inline Base.:+{order, dim}(S1::SymmetricTensor{order, dim}, S2::SymmetricTensor{order, dim}) = map(+, S1, S2)
+@inline Base.:-{order, dim}(S1::Tensor{order, dim}, S2::Tensor{order, dim}) = map(-, S1, S2)
+@inline Base.:-{order, dim}(S1::SymmetricTensor{order, dim}, S2::SymmetricTensor{order, dim}) = map(-, S1, S2)
+@inline Base.:-{order, dim}(S1::AnyTensor{order, dim}, S2::AnyTensor{order, dim}) =  ((SS1, SS2) = promote(S1, S2); map(-, SS1, SS2))
+@inline Base.:+{order, dim}(S1::AnyTensor{order, dim}, S2::AnyTensor{order, dim}) =  ((SS1, SS2) = promote(S1, S2); map(+, SS1, SS2))
 
-@inline Base.:+(S1::AbstractTensor, S2::AbstractTensor) = throw(DimensionMismatch("dimension and order must match"))
-@inline Base.:-(S1::AbstractTensor, S2::AbstractTensor) = throw(DimensionMismatch("dimension and order must match"))
+@inline Base.:*(S::AbstractTensor, n::Number) = map(@inline(function(x) x*n; end), S)
+@inline Base.:*(n::Number, S::AbstractTensor) = map(@inline(function(x) n*x; end), S)
+@inline Base.:/(S::AbstractTensor, n::Number) = map(@inline(function(x) x/n; end), S)
 
-@inline Base.:*(S::AbstractTensor, n::Number) = map(x->(x*n), S)
-@inline Base.:*(n::Number, S::AbstractTensor) = map(x->(n*x), S)
-@inline Base.:/(S::AbstractTensor, n::Number) = map(x->(x/n), S)
+Base.:+(S1::AbstractTensor, S2::AbstractTensor) = throw(DimensionMismatch("dimension and order must match"))
+Base.:-(S1::AbstractTensor, S2::AbstractTensor) = throw(DimensionMismatch("dimension and order must match"))
 
 # map implementations
 @inline function Base.map{T <: AbstractTensor}(f, S::T)
     return apply_all(S, @inline function(i) @inboundsret f(S.data[i]); end)
 end
 
-@inline Base.map{order, dim, T1, T2}(f, S1::AbstractTensor{order, dim, T1}, S2::AbstractTensor{order, dim, T2}) = ((SS1, SS2) = promote(S1, S2); map(f, SS1, SS2))
+Base.map{order, dim}(f, S1::Tensor{order, dim}, S2::SymmetricTensor{order, dim}) = ((SS1, SS2) = promote(S1, S2); map(f, SS1, SS2))
+Base.map{order, dim}(f, S1::SymmetricTensor{order, dim}, S2::Tensor{order, dim}) = ((SS1, SS2) = promote(S1, S2); map(f, SS1, SS2))
 
-@inline function Base.map{T <: AllTensors}(f, S1::T, S2::T)
+@inline function Base.map(f, S1::AllTensors, S2::AllTensors)
```

Comment:
Hm, yeah. I sort of saw this as an implementation detail (that you would never call `map` yourself). Perhaps we call it `_map` or something? Or do we want to support the official `map`.